### PR TITLE
Have gArgs handle printing help

### PR DESCRIPTION
**Pull request by achow101**
Instead of each binary generating and printing it's own help string, have gArgs know what the args are and generate the help string.

This is the first commit of #13112 pulled out.

### CODE REVIEW

**Comment by jnewbery** [comment]
Concept ACK

**Comment by MarcoFalke** [comment]
I guess you'd have to cherry-pick the fixup to contrib/check-doc as well.

File: src/init.cpp [hunk 1]
```diff
@@ -332,197 +332,163 @@ static void OnRPCStopped()
...
+    gArgs.AddArg("-txindex", strprintf(_("Maintain a full transaction index, used by the getrawtransaction rpc call (default: %u)"), DEFAULT_TXINDEX), false, OptionsCategory::OPTIONS);
+
+    gArgs.AddArg("-addnode=<ip>", _("Add a node to connect to and attempt to keep the connection open (see the `addnode` RPC command help for more info)"), false, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-banscore=<n>", strprintf(_("Threshold for disconnecting misbehaving peers (default: %u)"), DEFAULT_BANSCORE_THRESHOLD), false, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-bantime=<n>", strprintf(_("Number of seconds to keep misbehaving peers from reconnecting (default: %u)"), DEFAULT_MISBEHAVING_BANTIME), false, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-bind=<addr>", _("Bind to given address and always listen on it. Use [host]:port notation for IPv6"), false, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-connect=<ip>", _("Connect only to the specified node(s, false, OptionsCategory::OPTIONS)); -connect=0 disables automatic connections (the rules for this peer are the same as for -addnode)"), false, OptionsCategory::CONNECTION);
```
Comment (MarcoFalke) [inline src/init.cpp]: Rebase issue?

**Review by maflcko — COMMENTED** [review]
Tested ACK 3333c7ef48cfde42619b7f6ae62893de192ed574:

* Bumped the man pages in a commit on current master, cherry-picked 3333c7ef48cfde42619b7f6ae62893de192ed574, Generated the man pages and checked that `git diff` only shows differences to to sort-by-name.
* Additionally checked that the strings are move-only by hand
* Did an utACK on the code changes.

File: src/init.cpp (hunk 1 above)
Comment (achow101) [reply src/init.cpp]: Oops, fixed

**Comment by achow101** [comment]
I added a commit for the check-doc fix.

**Comment by MarcoFalke** [comment]
re-ACK 3ea4d9080d615e09017c23e9678f47c9edddb869 (only changes were suggested fixups)

**Comment by promag** [comment]
utACK 3ea4d90.

Please squash `check-doc.py` fails for 1st commit.

**Comment by achow101** [comment]
Squashed

**Comment by sipa** [comment]
utACK dadad0d559f62ff7d9752655dc4bbb922e97423d

I did not review all the command line help/AddArg statements, but I did compare the output of `-help` and `-help -help-debug` before and after.

**Comment by MeshCollider** [comment]
Concept ACK, makes sense to split this out

**Comment by MarcoFalke** [comment]
Needs rebase

File: src/util.h [hunk 2]
```diff
@@ -229,6 +247,16 @@ class ArgsManager
...
+
+    /**
+     * Add argument
+     */
+    void AddArg(const std::string& name, const std::string& help, const bool debug_only, const OptionsCategory& cat);
+
+    /**
+     * Get the help string
+     */
+    std::string GetHelpMessage();
```
Comment (sipa) [inline src/util.h]: I believe this can be marked const.

File: src/util.cpp [hunk 3]
```diff
@@ -547,6 +547,55 @@ void ArgsManager::ForceSetArg(const std::string& strArg, const std::string& strV
...
+{
+    const bool show_debug = gArgs.GetBoolArg("-help-debug", false);
+
+    std::string usage = HelpMessageGroup(_("Options:"));
+
+    OptionsCategory last_cat = OptionsCategory::OPTIONS;
+    for (auto& arg : m_available_args) {
+        if (arg.first.first != last_cat) {
+            last_cat = arg.first.first;
+            if (last_cat == OptionsCategory::CONNECTION)
```
Comment (sipa) [inline src/util.cpp]: Code style nit: use braces when putting then/else branches on separate lines.

**Review by sipa — COMMENTED** [review]
reutACK 4d4185a4f0e40c033a587871839a47cb3f89ee93

Two nits below, but feel free to ignore. This is probably a pain to keep rebasing.

**Comment by MarcoFalke** [comment]
re-ACK 4d4185a4f0e40c033a587871839a47cb3f89ee93 (only change was rebase)

**Pull request merged**